operator/pkg/lbipam/lbipam.go

```diff
@@ -0,0 +1,1797 @@
+
+	// First sync all pools
+	poolSync := false
+	for !poolSync {
+		select {
+		case event := <-poolChan:
+			event.Handle(
+				func() error {
+					err := ipam.settleConflicts()
+					if err != nil {
+						ipam.logger.WithError(err).Error("Error while settling pool conflicts")
+						poolRetryQueue.Add(retry{fn: ipam.settleConflicts})
+					}
+
+					poolSync = true
+					return nil
+				},
+				func(k resource.Key, pool *cilium_api_v2alpha1.CiliumLoadBalancerIPPool) error {
+					err := ipam.handleNewPool(pool)
+					if err != nil {
+						ipam.logger.WithError(err).Error("Error while handling new pools")
+						poolRetryQueue.Add(retry{fn: func() error { return ipam.handleNewPool(pool) }})
+					}
+
+					return nil
+				},
+				func(k resource.Key, pool *cilium_api_v2alpha1.CiliumLoadBalancerIPPool) error {
+					ipam.logger.Error("Got a Pool deleted event while syncing")
+					return nil
+				},
+			)
+		}
+	}
+
+	// Retry any pool related errors before we start processing the services. If we start processing services while we
+	// are missing pools due to connectivity issues, we might remove ingress IPs and cause service disruption.
+
+	const retryTimeout = 5 * time.Second
+	const retryCount = 6
+	for poolRetryQueue.Len() > 0 {
+		// Wait a bit in case the error is transient
+		time.Sleep(retryTimeout)
+		// Retry every function in the queue once.
+		for i := 0; i < poolRetryQueue.Len(); i++ {
+			retryInt, _ := poolRetryQueue.Get()
+			retry := retryInt.(*retry)
+
+			if err := retry.fn(); err != nil {
+				retry.cnt++
+
+				// Give up after a while. Better to crash the operator than to corrupt the presumably still working
+				// service state.
+				if retry.cnt > retryCount {
+					return fmt.Errorf("Initialization failed: %w", err)
+				}
+
+				// Requeue if the error still isn't resolved.
+				poolRetryQueue.Add(retry)
+			}
+		}
+	}
+
+	// Then sync all services
+	svcSync := false
+	for !svcSync {
+		select {
+		case event := <-svcChan:
+			event.Handle(
+				func() error {
+					svcSync = true
+
+					err := ipam.satisfyServices()
+					if err != nil {
+						// Let the resource retry after the initialization
+						return err
+					}
+
+					err = ipam.updateAllPoolCounts()
+					if err != nil {
+						// Let the resource retry after the initialization
+						return err
+					}
+
+					return nil
+				},
+				func(k resource.Key, svc *slim_core_v1.Service) error {
+					err := ipam.handleUpsertService(svc)
+					if err != nil {
+						// Let the resource retry after the initialization
+						return err
+					}
+
+					return nil
+				},
+				func(k resource.Key, svc *slim_core_v1.Service) error {
+					ipam.logger.Error("Got a Service deleted event while syncing")
+					return nil
+				},
+			)
+		}
+	}
+
+	return nil
+}
+
+func (ipam *LBIPAM) poolOnUpsert(k resource.Key, pool *cilium_api_v2alpha1.CiliumLoadBalancerIPPool) error {
+	// Deep copy so we get a version we are allowed to update
+	pool = pool.DeepCopy()
```

Comment:
From what I can see we are not modifying the `pool` object so we don't need to perform DeepCopy.

Reply:
We modify the pool's status, since we don't store the pool status seperately we have to deep copy the whole thing so we can take ownership and don't have to worry about changing a shared copy owned by the store.